### PR TITLE
Added Zip operation and Sample operation

### PR DESCRIPTION
**Pull request by Xintong-Zhan**
Files modified:
* examples/filter_even.cpp: Added usage examples of Sample operation.
* examples/simple.cpp: Added a usage eample of Zip operation.
* include/filter_rdd.h: Included ramdom header for the support of Sample operation.

Files added:
* include/sample_rdd.h: Sample operation. This operation is based on FilterView by passing random boolean function as a param to FilterView.
* include/zipped_rdd.h: Zip operation. No pipeline is implemented for the Zip operation since both params are Rdd, and using pipeline with one Rdd to be function param is a little bit weird.
* tests/sample_rdd_test.cpp: Google test for Sample.
* tests/zipped_rdd_test.cpp: Goole test for Zip.

### CODE REVIEW

File: include/filter_rdd.h
Comment (Alex-XJK) [inline include/filter_rdd.h:5]: Should we add the <random> to "sample_rdd.h" or here? Since it seems the filter_rdd itself does not need the random library?

File: include/sample_rdd.h
Comment (Alex-XJK) [inline include/sample_rdd.h:10]: I would suggest adding the doxygen-required comments since the main documentation has already been deployed, later commited file should handle this before the merge. 
```suggestion

/** @ingroup g_trans
 *  @defgroup t_Sample Sample
 *  This forms the Sample Transformation of our cpark library
 *  @image html sample.drawio.png "SampleRdd diagram" width=50%
 *  @image latex sample.drawio.png "SampleRdd diagram" width=10cm
 *  @see SampleRdd
 *  @see Sample
 *  @{
 */
```

File: include/sample_rdd.h
Comment (Alex-XJK) [inline include/sample_rdd.h:89]:
```suggestion

/** @} */ // end of t_Sample
```

File: include/zipped_rdd.h
Comment (Alex-XJK) [inline include/zipped_rdd.h:12]: The same requirements go here.
```suggestion

/** @ingroup g_trans
 *  @defgroup t_Zip Zip
 *  This forms the Zip Transformation of our cpark library
 *  @image html zip.drawio.png "ZippedRdd diagram" width=50%
 *  @image latex zip.drawio.png "ZippedRdd diagram" width=10cm
 *  @see ZippedRdd
 *  @{
 */
```

File: include/zipped_rdd.h
Comment (Alex-XJK) [inline include/zipped_rdd.h:62]:
```suggestion
/** @} */ // end of t_Zip

}  // namespace cpark
```

File: include/zipped_rdd.h
Comment (Alex-XJK) [inline include/zipped_rdd.h:27]: Shall we unify the name convention? Currently, it looks like some Rdd's are named in the present tense (Sample, Merge), and some are named in the past tense (Zipped, Transformed).

**Review by Alex-XJK — CHANGES_REQUESTED** [review]
I think the doxygen comment should be added. I have already provide some skeleton code for you, please double check it and add additional description if needed.

**Comment by Alex-XJK** [comment]
> ## [Codecov](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU) Report
> Attention: `2 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`b15424d`)](https://app.codecov.io/gh/Alex-XJK/cpark/commit/b15424d18c37074d8ccb7c5822e70acc455d2d77?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU) 97.29% compared to head [(`908a0bf`)](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU) 97.22%.
> > Report is 2 commits behind head on main.
> 
> [Files](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU)	Patch %	Lines
> [include/zipped_rdd.h](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU#diff-aW5jbHVkZS96aXBwZWRfcmRkLmg=)	92.30%	[1 Missing ⚠️ ](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU)
> [tests/sample_rdd_test.cpp](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU#diff-dGVzdHMvc2FtcGxlX3JkZF90ZXN0LmNwcA==)	95.00%	[1 Missing ⚠️ ](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU).

It is strange that one of the lines in the TEST script didn't get covered, because they are the ones we used to test other headers.  
If there are no other coding issues hidden behind, I think we can simply ignore this coverage issue because, obviously, getting it functionally correct is much more important than getting it 100% coverage, at least for now.  
If other reviewers approved all other parts, but this Codecov is still blocking the merge, let me know, and I can force merge it 😊

File: include/sample_rdd.h
Comment (chenlighten) [inline include/sample_rdd.h:55]: This would constrain the element type of the previous RDD to be merely `int`. The `int i` argument is only a place holder to fit the requirements of `FilterView`'s function argument, and is not actually used, isn't it?

 So I would suggest change it to `std::function<bool([int](utils::RddElementType<R>))> func = [this](utils::RddElementType<R>) ...`

File: include/sample_rdd.h
Comment (chenlighten) [inline include/sample_rdd.h:18]: This requirement would constrain the element type of the previous RDD to be merely `int`. I would suggest removing this requires clause. Please also see the comment in line 55.

File: include/sample_rdd.h
Comment (chenlighten) [inline include/sample_rdd.h:29]: How about also documenting what will happen if `fraction` does not fall in range `[0,1]`?

File: include/zipped_rdd.h
Comment (chenlighten) [reply include/zipped_rdd.h:27]: I think it is a relatively minor issue. But if any of you are available, it does not hurt to rename them to past tense.

File: include/zipped_rdd.h
Comment (chenlighten) [inline include/zipped_rdd.h:37]: Seems that `cnt` is not used in the lambda expression?

**Review by chenlighten — COMMENTED** [review]
Thanks for finishing `sampe_rdd` and `zipped_rdd`. Commented some minor issues. You can take a look of them if you have time. If you are busy, feel free to take care of them in the future.

File: include/filter_rdd.h
Comment (Xintong-Zhan) [reply include/filter_rdd.h:5]: Oh yes you're right. I didn't pay attention to that and I will modify it. Thanks

File: include/sample_rdd.h
Comment (Xintong-Zhan) [reply include/sample_rdd.h:10]: Got it. Will refer to the existing main repo to add the comments

**Comment by Xintong-Zhan** [comment]
> ## [Codecov](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU) Report
> Attention: `2 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`b15424d`)](https://app.codecov.io/gh/Alex-XJK/cpark/commit/b15424d18c37074d8ccb7c5822e70acc455d2d77?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU) 97.29% compared to head [(`908a0bf`)](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU) 97.22%.
> > Report is 2 commits behind head on main.
> 
> [Files](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU)	Patch %	Lines
> [include/zipped_rdd.h](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU#diff-aW5jbHVkZS96aXBwZWRfcmRkLmg=)	92.30%	[1 Missing ⚠️ ](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU)
> [tests/sample_rdd_test.cpp](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU#diff-dGVzdHMvc2FtcGxlX3JkZF90ZXN0LmNwcA==)	95.00%	[1 Missing ⚠️ ](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/Alex-XJK/cpark/pull/26?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Alex_XU).

For the Google Test script, the line that didn't get covered is for the SampleRdd operation with fraction = 0, which means the returned Rdd should contain nothing, thus the program won't get in the loop body for Rdd and I guess that's the reason for not being covered.   And because of the random feature, the only deterministic case I could test in Google test script is fraction = 1 and fraction = 0. For other random sample cases, I present them in the ./examples directory.

File: include/sample_rdd.h
Comment (Xintong-Zhan) [reply include/sample_rdd.h:89]: Got it. Will refer to the existing main repo to add the comments.

File: include/zipped_rdd.h
Comment (Xintong-Zhan) [reply include/zipped_rdd.h:12]: Got it. Will refer to the existing main repo to add the comments

File: include/zipped_rdd.h
Comment (Xintong-Zhan) [reply include/zipped_rdd.h:62]: Got it. Will refer to the existing main repo to add the comments

File: include/sample_rdd.h
Comment (Alex-XJK) [reply include/sample_rdd.h:10]: Thanks for your reply. 
Yes, providing a more detailed text description is welcomed. 
However, to ease your work, I have already provided the most basic solution in the reviews. If you want, you can just click Commit Suggestion on GitHub.

File: tests/sample_rdd_test.cpp
Comment (Alex-XJK) [inline tests/sample_rdd_test.cpp:30]: This is a follow up to your https://github.com/Alex-XJK/cpark/pull/26#issuecomment-1839973690

```suggestion
       EXPECT_TRUE(split.begin() == split.end());
```

**Review by Alex-XJK — APPROVED** [review]
Thank you for your effort!

**Pull request merged**
